Extensions/3D/Model3DRuntimeObject.ts

```diff
@@ -118,32 +119,54 @@ namespace gdjs {
       this._materialType = this._convertMaterialType(
         objectData.content.materialType
       );
-      this._updateModel(objectData);
+
+      this.onChangeModelResourceName(objectData);
+
+      // *ALWAYS* call `this.onCreated()` at the very end of your object constructor.
+      this.onCreated();
+    }
+
+    private onChangeModelResourceName(objectData) {
+      this._renderer._reloadModel(this, this._runtimeScene, objectData);
       if (this._animations.length > 0) {
         this._renderer.playAnimation(
           this._animations[0].source,
           this._animations[0].loop
         );
       }
-
-      // *ALWAYS* call `this.onCreated()` at the very end of your object constructor.
-      this.onCreated();
     }
 
     updateFromObjectData(
       oldObjectData: Model3DObjectData,
       newObjectData: Model3DObjectData
     ): boolean {
       super.updateFromObjectData(oldObjectData, newObjectData);
+
       if (
+        oldObjectData.content.materialType !==
+        newObjectData.content.materialType
+      ) {
+        this._materialType = this._convertMaterialType(
+          newObjectData.content.materialType
+        );
+      }
```

Comment:
Since _updateModel is expensive (clones the entire THREE model tree, traverses it, etc), I've gone ahead and done the optimization of not calling it twice here while I was at it. I'll undo it if you think it might break something.